packages/server-reason-react-ppx/server_reason_react_ppx.ml

```diff
@@ -854,7 +856,213 @@ let rewrite_jsx =
+        let of_json = make_of_json ~loc core_type [%expr args.([%e eint ~loc i])] in
+        value_binding ~loc
+          ~pat:[%pat? [%p ppat_var ~loc { txt = label; loc }]]
+          ~expr:
+            [%expr
+              try [%e of_json]
+              with _ ->
+                failwith
+                  (Printf.sprintf "server-reason-react: error on decoding argument '%s'. EXPECTED: %s, RECEIVED: %s"
+                     [%e estring ~loc label] [%e estring ~loc core_type_string]
+                     (args.([%e eint ~loc i]) |> Yojson.Basic.to_string))])
+      args_to_decode
+
+  let create_router_handler ~loc ~id ~function_name ~args ~core_type =
+    let apply_args = map_arguments_to_expressions ~loc args in
+    let response_expr = pexp_apply ~loc [%expr [%e evar ~loc function_name].call] apply_args in
+
+    let encoded_response_expr = encode_function_response ~loc ~response_expr ~core_type in
+    let args_to_decode =
+      List.filter_map
+        ~f:(fun arg ->
+          match arg with
+          | Ok (arg_label, Some arg_name, core_type) -> Some (arg_label, arg_name, core_type)
+          | Ok _ -> None
+          | Error _ -> None)
+        args
+    in
+
+    let body_expr =
+      match args_to_decode with
+      | [] -> encoded_response_expr
+      | args_to_decode ->
+          let decoded_expr = decode_arguments_vb ~loc args_to_decode in
+          pexp_let ~loc Nonrecursive decoded_expr encoded_response_expr
+    in
+    [%stri FunctionReferences.register [%e estring ~loc id] (Body (fun args -> [%e body_expr]))]
+
+  let create_server_function_contract ~loc id expression =
+    let fn = [%expr { Runtime.id = [%e estring ~loc id]; call = [%e expression] }] in
+    fn
+
+  let transform_native_function structure_item =
+    match structure_item.pstr_desc with
+    | Pstr_value (rec_flag, value_bindings) when isReactServerFunctionBinding (List.hd value_bindings) ->
+        let loc = structure_item.pstr_loc in
+        let vb = List.hd value_bindings in
+
+        let function_name = get_function_name vb in
+        let args = get_arguments vb.pvb_expr |> List.map ~f:get_arg_details |> List.rev in
+        let base_fn = vb.pvb_expr in
+        let return_core_type = get_response_type base_fn in
+        let id = generate_id ~loc:vb.pvb_loc function_name in
+        let value_binding = { vb with pvb_expr = create_server_function_contract ~loc:vb.pvb_loc id base_fn } in
+        let stri =
+          [%stri
+            include struct
+              [%%i pstr_value ~loc rec_flag [ value_binding ]]
+              [%%i create_router_handler ~loc ~id ~function_name ~args ~core_type:return_core_type]
+            end]
+        in
+        stri
+    | _ -> structure_item
+
+  let create_client_function ~loc id args =
+    let apply_args = map_arguments_to_expressions ~loc args |> List.map ~f:(fun (_, expr) -> (Nolabel, expr)) in
+    let fn =
+      [%expr
+        (* TODO: support others bundlers? *)
```

Comment:
Currently we only support our esbuild, and there's no plans on supporting Webpack at the moment (or any other really)